File: scenarios/moderate/mobile/flutter/lib/custom_dropdown_navigation.dart
import 'package:flutter/material.dart';
import 'package:flutter/semantics.dart';

class Product {
  final String name;
  final String category;

  Product({required this.name, required this.category});
}

class Option {
  final String value;
  final String text;

  Option({required this.value, required this.text});
}

class CustomDropdownNavigation extends StatefulWidget {
  const CustomDropdownNavigation({Key? key}) : super(key: key);

  @override
  State<CustomDropdownNavigation> createState() => _CustomDropdownNavigationState();
}

class _CustomDropdownNavigationState extends State<CustomDropdownNavigation> {
  bool isOpen = false;
  String selectedValue = 'all';
  String selectedText = 'All Categories';
  int currentIndex = 0;

  final List<Option> options = [
    Option(value: 'all', text: 'All Categories'),
    Option(value: 'electronics', text: 'Electronics'),
    Option(value: 'clothing', text: 'Clothing'),
    Option(value: 'books', text: 'Books'),
    Option(value: 'home', text: 'Home & Garden'),
  ];

  final Map<String, List<Product>> products = {
    'all': [
      Product(name: 'iPhone 15', category: 'Electronics'),
      Product(name: 'MacBook Pro', category: 'Electronics'),
      Product(name: 'Cotton T-Shirt', category: 'Clothing'),
      Product(name: 'JavaScript Guide', category: 'Books'),
      Product(name: 'Garden Tools', category: 'Home & Garden'),
    ],
    'electronics': [
      Product(name: 'iPhone 15', category: 'Electronics'),
      Product(name: 'MacBook Pro', category: 'Electronics'),
    ],
    'clothing': [
      Product(name: 'Cotton T-Shirt', category: 'Clothing'),
    ],
    'books': [
      Product(name: 'JavaScript Guide', category: 'Books'),
    ],
    'home': [
      Product(name: 'Garden Tools', category: 'Home & Garden'),
    ],
  };

  List<Product> get filteredProducts => products[selectedValue] ?? [];

  void toggleDropdown() {
    setState(() {
      isOpen = !isOpen;
    });
  }

  void selectOption(String value, String text, int index) {
    setState(() {
      selectedValue = value;
      selectedText = text;
      currentIndex = index;
      isOpen = false;
    });

    SemanticsService.announce('Selected $text', TextDirection.ltr);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: const Color(0xFFF5F5F5),
      body: SingleChildScrollView(
        padding: const EdgeInsets.all(20),
        child: Column(
          children: [
            // Main Content Card
            Card(
              elevation: 4,
              shape: RoundedRectangleBorder(
                borderRadius: BorderRadius.circular(8),
              ),
              child: Padding(
                padding: const EdgeInsets.all(30),
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    // Title
                    const Text(
                      'Product Filter',
                      style: TextStyle(
                        fontSize: 24,
                        fontWeight: FontWeight.bold,
                        color: Color(0xFF333333),
                      ),
                    ),
                    const SizedBox(height: 16),
                    
                    // Description
                    const Text(
                      'Filter products by category using the dropdown below.',
                      style: TextStyle(
                        fontSize: 16,
                        color: Color(0xFF666666),
                      ),
                    ),
                    const SizedBox(height: 30),
                    
                    // Filter Section
                    Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        const Text(
                          'Filter by Category:',
                          style: TextStyle(
                            fontSize: 18,
                            fontWeight: FontWeight.bold,
                            color: Color(0xFF333333),
                          ),
                        ),
                        const SizedBox(height: 8),
                        
                        // Custom Dropdown
                        SizedBox(
                          width: 200,
                          child: Stack(
                            children: [
                              // Dropdown Button
                              Semantics(
                                label: 'Category filter, $selectedText, double tap to open dropdown',
                                hint: 'Opens category selection dropdown',
                                button: true,
                                child: Material(
                                  color: Colors.transparent,
                                  child: InkWell(
                                    onTap: toggleDropdown,
                                    borderRadius: BorderRadius.circular(isOpen ? 8 : 8),
                                    child: Container(
                                      height: 44,
                                      decoration: BoxDecoration(
                                        color: const Color(0xFFF8F9FA),
                                        border: Border.all(
                                          color: isOpen 
                                              ? const Color(0xFF007BFF) 
                                              : const Color(0xFFDDDDDD),
                                          width: 2,
                                        ),
                                        borderRadius: BorderRadius.circular(isOpen ? 8 : 8),
                                      ),
                                      child: Padding(
                                        padding: const EdgeInsets.symmetric(horizontal: 16),
                                        child: Row(
                                          mainAxisAlignment: MainAxisAlignment.spaceBetween,
                                          children: [
                                            Text(
                                              selectedText,
                                              style: const TextStyle(
                                                fontSize: 16,
                                                color: Color(0xFF333333),
                                              ),
                                            ),
                                            AnimatedRotation(
                                              turns: isOpen ? 0.5 : 0,
                                              duration: const Duration(milliseconds: 200),
                                              child: const Icon(
                                                Icons.keyboard_arrow_down,
                                                color: Color(0xFF666666),
                                              ),
                                            ),
                                          ],
                                        ),
                                      ),
                                    ),
                                  ),
                                ),
                              ),
                              
                              // Dropdown Menu
                              if (isOpen)
                                Positioned(
                                  top: 44,
                                  left: 0,
                                  right: 0,
                                  child: Material(
                                    elevation: 8,
                                    borderRadius: const BorderRadius.only(
                                      bottomLeft: Radius.circular(8),
                                      bottomRight: Radius.circular(8),
                                    ),
                                    child: Container(
                                      decoration: BoxDecoration(
                                        color: Colors.white,
                                        border: Border.all(
                                          color: const Color(0xFF007BFF),
                                          width: 2,
                                        ),
                                        borderRadius: const BorderRadius.only(
                                          bottomLeft: Radius.circular(8),
                                          bottomRight: Radius.circular(8),
                                        ),
                                      ),
                                      child: Column(
                                        children: options.asMap().entries.map((entry) {
                                          int index = entry.key;
                                          Option option = entry.value;
                                          bool isSelected = selectedValue == option.value;
                                          
                                          return Semantics(
                                            label: option.text,
                                            selected: isSelected,
                                            button: true,
                                            child: Material(
                                              color: Colors.transparent,
                                              child: InkWell(
                                                onTap: () => selectOption(option.value, option.text, index),
                                                child: Container(
                                                  padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 12),
                                                  decoration: BoxDecoration(
                                                    color: isSelected 
                                                        ? const Color(0xFF007BFF) 
                                                        : Colors.transparent,
                                                    border: index < options.length - 1
                                                        ? const Border(
                                                            bottom: BorderSide(
                                                              color: Color(0xFFEEEEEE),
                                                              width: 1,
                                                            ),
                                                          )
                                                        : null,
                                                  ),
                                                  child: Row(
                                                    mainAxisAlignment: MainAxisAlignment.spaceBetween,
                                                    children: [
                                                      Text(
                                                        option.text,
                                                        style: TextStyle(
                                                          fontSize: 16,
                                                          color: isSelected 
                                                              ? Colors.white 
                                                              : const Color(0xFF333333),
                                                        ),
                                                      ),
                                                      if (isSelected)
                                                        const Icon(
                                                          Icons.check,
                                                          color: Colors.white,
                                                          size: 20,
                                                        ),
                                                    ],
                                                  ),
                                                ),
                                              ),
                                            ),
                                          );
                                        }).toList(),
                                      ),
                                    ),
                                  ),
                                ),
                            ],
                          ),
                        ),
                      ],
                    ),
                    const SizedBox(height: 30),
                    
                    // Results Section
                    Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        const Text(
                          'Filtered Results',
                          style: TextStyle(
                            fontSize: 20,
                            fontWeight: FontWeight.bold,
                            color: Color(0xFF333333),
                          ),
                        ),
                        const SizedBox(height: 16),
                        
                        SizedBox(
                          height: 200,
                          child: GridView.builder(
                            gridDelegate: const SliverGridDelegateWithFixedCrossAxisCount(
                              crossAxisCount: 2,
                              childAspectRatio: 1.5,
                              crossAxisSpacing: 16,
                              mainAxisSpacing: 16,
                            ),
                            itemCount: filteredProducts.length,
                            itemBuilder: (context, index) {
                              final product = filteredProducts[index];
                              return Semantics(
                                label: '${product.name}, ${product.category}',
                                child: Card(
                                  elevation: 2,
                                  shape: RoundedRectangleBorder(
                                    borderRadius: BorderRadius.circular(8),
                                  ),
                                  child: Padding(
                                    padding: const EdgeInsets.all(20),
                                    child: Column(
                                      mainAxisAlignment: MainAxisAlignment.center,
                                      children: [
                                        Text(
                                          product.name,
                                          style: const TextStyle(
                                            fontSize: 16,
                                            fontWeight: FontWeight.w600,
                                            color: Color(0xFF333333),
                                          ),
                                          textAlign: TextAlign.center,
                                        ),
                                        const SizedBox(height: 8),
                                        Text(
                                          product.category,
                                          style: const TextStyle(
                                            fontSize: 12,
                                            color: Color(0xFF666666),
                                          ),
                                          textAlign: TextAlign.center,
                                        ),
                                      ],
                                    ),
                                  ),
                                ),
                              );
                            },
                          ),
                        ),
                      ],
                    ),
                  ],
                ),
              ),
            ),
            const SizedBox(height: 20),
            
          ],
        ),
      ),
    );
  }
}

File: scenarios/moderate/mobile/android/compose/CustomDropdownNavigation.kt
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ArrowDropDown
import androidx.compose.material.icons.filled.ArrowDropUp
import androidx.compose.material.icons.filled.Check
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
data class Product(
    val name: String,
    val category: String
)
data class Option(
    val value: String,
    val text: String
)
@OptIn(ExperimentalMaterial3Api::class)
@Composable
    val options = listOf(
        Option("all", "All Categories"),
        Option("electronics", "Electronics"),
        Option("clothing", "Clothing"),
        Option("books", "Books"),
        Option("home", "Home & Garden")
    )
    val products = mapOf(
        "all" to listOf(
            Product("iPhone 15", "Electronics"),
            Product("MacBook Pro", "Electronics"),
            Product("Cotton T-Shirt", "Clothing"),
            Product("JavaScript Guide", "Books"),
            Product("Garden Tools", "Home & Garden")
        ),
        "electronics" to listOf(
            Product("iPhone 15", "Electronics"),
            Product("MacBook Pro", "Electronics")
        ),
        "clothing" to listOf(
            Product("Cotton T-Shirt", "Clothing")
        ),
        "books" to listOf(
            Product("JavaScript Guide", "Books")
        ),
        "home" to listOf(
            Product("Garden Tools", "Home & Garden")
        )
    )
    val filteredProducts = products[selectedValue] ?: emptyList()
    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(Color(0xFFF5F5F5))
            .padding(20.dp),
        verticalArrangement = Arrangement.spacedBy(20.dp)
    ) {
        Card(
            modifier = Modifier.fillMaxWidth(),
            colors = CardDefaults.cardColors(containerColor = Color.White),
            elevation = CardDefaults.cardElevation(defaultElevation = 4.dp)
        ) {
            Column(
                modifier = Modifier.padding(30.dp),
                verticalArrangement = Arrangement.spacedBy(20.dp)
            ) {
                Text(
                    text = "Product Filter",
                    fontSize = 24.sp,
                    fontWeight = FontWeight.Bold,
                    color = Color(0xFF333333)
                )
                Text(
                    fontSize = 16.sp,
                    color = Color(0xFF666666)
                )
                Column(
                    verticalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    Text(
                        text = "Filter by Category:",
                        fontSize = 18.sp,
                        fontWeight = FontWeight.Bold,
                        color = Color(0xFF333333)
                    )
                    // Custom Dropdown Implementation
                    Box {
                        Button(
                            onClick = { isOpen = !isOpen },
                            modifier = Modifier
                                .fillMaxWidth(0.5f)
                                .height(44.dp)
                                    contentDescription = "Category filter, $selectedText, double tap to open dropdown"
                                },
                            colors = ButtonDefaults.buttonColors(
                                containerColor = Color(0xFFF8F9FA)
                            ),
                            shape = RoundedCornerShape(8.dp),
                            border = androidx.compose.foundation.BorderStroke(
                                2.dp,
                                if (isOpen) Color(0xFF007BFF) else Color(0xFFDDDDDD)
                            )
                        ) {
                            Row(
                                modifier = Modifier.fillMaxWidth(),
                                horizontalArrangement = Arrangement.SpaceBetween,
                                verticalAlignment = Alignment.CenterVertically
                            ) {
                                Text(
                                    text = selectedText,
                                    color = Color(0xFF333333),
                                    fontSize = 16.sp
                                )
                                Icon(
                                    imageVector = if (isOpen) Icons.Default.ArrowDropUp else Icons.Default.ArrowDropDown,
                                    contentDescription = if (isOpen) "Close dropdown" else "Open dropdown",
                                    tint = Color(0xFF666666)
                                )
                            }
                        }
                        // Dropdown Options
                        if (isOpen) {
                            Card(
                                modifier = Modifier
                                    .fillMaxWidth(0.5f)
                                    .offset(y = 44.dp),
                                colors = CardDefaults.cardColors(containerColor = Color.White),
                                elevation = CardDefaults.cardElevation(defaultElevation = 8.dp),
                                shape = RoundedCornerShape(bottomStart = 8.dp, bottomEnd = 8.dp)
                            ) {
                                Column {
                                    options.forEachIndexed { index, option ->
                                        Row(
                                            modifier = Modifier
                                                .fillMaxWidth()
                                                .clickable {
                                                    selectOption(
                                                        value = option.value,
                                                        text = option.text,
                                                        index = index,
                                                        selectedValue = { selectedValue = it },
                                                        selectedText = { selectedText = it },
                                                        currentIndex = { currentIndex = it },
                                                        isOpen = { isOpen = it }
                                                    )
                                                }
                                                .padding(12.dp)
                                                .background(
                                                    if (selectedValue == option.value) Color(0xFF007BFF) else Color.Transparent
                                                )
                                                    contentDescription = option.text
                                                    if (selectedValue == option.value) {
                                                        selected()
                                                    }
                                                },
                                            horizontalArrangement = Arrangement.SpaceBetween,
                                            verticalAlignment = Alignment.CenterVertically
                                        ) {
                                            Text(
                                                text = option.text,
                                                color = if (selectedValue == option.value) Color.White else Color(0xFF333333),
                                                fontSize = 16.sp
                                            )
                                            if (selectedValue == option.value) {
                                                Icon(
                                                    imageVector = Icons.Default.Check,
                                                    contentDescription = "Selected",
                                                    tint = Color.White
                                                )
                                            }
                                        }
                                        if (index < options.size - 1) {
                                            Divider(
                                                color = Color(0xFFEEEEEE),
                                                thickness = 1.dp
                                            )
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
                Column(
                    verticalArrangement = Arrangement.spacedBy(16.dp)
                ) {
                    Text(
                        text = "Filtered Results",
                        fontSize = 20.sp,
                        fontWeight = FontWeight.Bold,
                        color = Color(0xFF333333)
                    )
                    LazyVerticalGrid(
                        columns = GridCells.Adaptive(200.dp),
                        horizontalArrangement = Arrangement.spacedBy(16.dp),
                        verticalArrangement = Arrangement.spacedBy(16.dp),
                        modifier = Modifier.height(200.dp)
                    ) {
                        items(filteredProducts) { product ->
                            Card(
                                modifier = Modifier.fillMaxWidth(),
                                colors = CardDefaults.cardColors(containerColor = Color.White),
                                elevation = CardDefaults.cardElevation(defaultElevation = 2.dp),
                                shape = RoundedCornerShape(8.dp)
                            ) {
                                Column(
                                    modifier = Modifier.padding(20.dp),
                                    horizontalAlignment = Alignment.CenterHorizontally,
                                    verticalArrangement = Arrangement.spacedBy(8.dp)
                                ) {
                                    Text(
                                        text = product.name,
                                        fontSize = 16.sp,
                                        fontWeight = FontWeight.SemiBold,
                                        color = Color(0xFF333333),
                                        textAlign = TextAlign.Center
                                    )
                                    Text(
                                        text = product.category,
                                        fontSize = 12.sp,
                                        color = Color(0xFF666666),
                                        textAlign = TextAlign.Center
                                    )
                                }
                            }
                        }
                    }
                }
            }
        }
        Card(
            modifier = Modifier.fillMaxWidth(),
            colors = CardDefaults.cardColors(containerColor = Color(0xFFE9ECEF)),
            shape = RoundedCornerShape(8.dp)
        ) {
            Column(
                modifier = Modifier.padding(20.dp),
                verticalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                Text(
                }
                Text(
                    text = "How to Fix:",
                    fontSize = 18.sp,
                    fontWeight = FontWeight.Bold,
                    color = Color(0xFF333333)
                )
                Column(
                    verticalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    fixItem("Handle Escape key to close dropdown")
                }
            }
        }
    }
}
@Composable
    Row(
        verticalAlignment = Alignment.Top,
        horizontalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Text(
            text = "•",
            fontWeight = FontWeight.Bold,
            color = Color(0xFF333333)
        )
        Text(
            text = text,
            fontSize = 14.sp,
            color = Color(0xFF333333)
        )
    }
}
@Composable
private fun fixItem(text: String) {
    Row(
        verticalAlignment = Alignment.Top,
        horizontalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Text(
            text = "•",
            fontWeight = FontWeight.Bold,
            color = Color(0xFF333333)
        )
        Text(
            text = text,
            fontSize = 14.sp,
            color = Color(0xFF333333)
        )
    }
}
private fun selectOption(
    value: String,
    text: String,
    index: Int,
    selectedValue: (String) -> Unit,
    selectedText: (String) -> Unit,
    currentIndex: (Int) -> Unit,
    isOpen: (Boolean) -> Unit
) {
    selectedValue(value)
    selectedText(text)
    currentIndex(index)
    isOpen(false)
}